fft/src/KokkosFFT_FFTW_Types.hpp

```diff
@@ -0,0 +1,154 @@
+  static constexpr FFTWTransformType m_type = std::is_same_v<T1, float>
+                                                  ? FFTWTransformType::C2C
+                                                  : FFTWTransformType::Z2Z;
+  static constexpr FFTWTransformType type() { return m_type; };
+};
+
+/// \brief A class that wraps fftw_plan and fftwf_plan for RAII
+template <typename ExecutionSpace, typename T1, typename T2>
+struct ScopedFFTWPlan {
+ private:
+  using floating_point_type = KokkosFFT::Impl::base_floating_point_type<T1>;
+  using plan_type =
+      std::conditional_t<std::is_same_v<floating_point_type, float>, fftwf_plan,
+                         fftw_plan>;
+  plan_type m_plan;
+  bool m_is_created = false;
+
+ public:
+  template <typename InScalarType, typename OutScalarType>
+  ScopedFFTWPlan(const ExecutionSpace &exec_space, int rank, const int *n,
+                 int howmany, InScalarType *in, const int *inembed, int istride,
+                 int idist, OutScalarType *out, const int *onembed, int ostride,
+                 int odist, [[maybe_unused]] int sign, unsigned flags) {
+    init_threads(exec_space);
+    constexpr auto type = fftw_transform_type<ExecutionSpace, T1, T2>::type();
+    if constexpr (type == KokkosFFT::Impl::FFTWTransformType::R2C) {
+      m_plan =
+          fftwf_plan_many_dft_r2c(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::D2Z) {
+      m_plan =
+          fftw_plan_many_dft_r2c(rank, n, howmany, in, inembed, istride, idist,
+                                 out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::C2R) {
+      m_plan =
+          fftwf_plan_many_dft_c2r(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::Z2D) {
+      m_plan =
+          fftw_plan_many_dft_c2r(rank, n, howmany, in, inembed, istride, idist,
+                                 out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::C2C) {
+      m_plan =
+          fftwf_plan_many_dft(rank, n, howmany, in, inembed, istride, idist,
+                              out, onembed, ostride, odist, sign, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::Z2Z) {
+      m_plan = fftw_plan_many_dft(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, sign, flags);
+    }
+    m_is_created = true;
+  }
+  ~ScopedFFTWPlan() noexcept {
+    cleanup_threads();
+    if constexpr (std::is_same_v<plan_type, fftwf_plan>) {
+      if (m_is_created) fftwf_destroy_plan(m_plan);
+    } else {
+      if (m_is_created) fftw_destroy_plan(m_plan);
+    }
```

Comment:
In which situation do we end up with `m_is_created == false` ?

Reply:
I may expect that `fftw_plan_many_dft` may fail if we request too much memory

Reply:
That is a possibility but I am not sure to understand how we capture this, I don't any error status variable in this snippet ?

Reply:
If I understand correctly, the code will abort if something gets wrong with `fftw`. So may be this check is not useful